scipy/optimize/optimize.py

```diff
@@ -3061,6 +2912,576 @@ def show_options(solver=None, method=None, disp=True):
+        Important attributes are: `x` the solution array, `success` a Boolean
+        flag indicating if the optimizer finished successfully and `message`
+        which describes the cause of the termination. See `OptimizeResult`
+        for a description of other attributes.
+        """
+        # override to set other fields.
+        if self.nit == 0:
+            return None
+
+        result = OptimizeResult(
+            x=self.x,
+            fun=self.fun,
+            nfev=self.nfev,
+            nit=self.nit,
+            message=self.message,
+            success=(self.warning_flag == 0 and self.converged()))
+        return result
+
+    def __call__(self, iterations):
+        """
+        Advance the solver by a number of steps.
+
+        Parameters
+        ----------
+        iterations: int
+            The number of iterations to perform
+
+        Returns
+        -------
+        result: OptimizeResult
+            The optimization result represented as a OptimizeResult object.
+        """
+        # TODO interupt once convergence is reached?
+        for it in range(iterations):
+            self.x, self.fun = next(self)
+
+        self._finish_up()
+        return self.result
+
+    @property
+    def N(self):
+        """
+        The dimensionality of the problem, `np.size(x)`.
+        """
+        if self.x is not None:
+            return self.x.size
+        elif self.x0 is not None:
+            return self.x0.size
+        return RuntimeError("Cannot determine problem size at this time,"
+                            " perform at least one iteration")
+
+    def solve(self):
+        """
+        Run the solver through to completion.
+
+        Returns
+        -------
+        result: OptimizeResult
+            The optimization result represented as a `OptimizeResult` object.
+            Important attributes are: `x` the solution array, `success` a
+            Boolean flag indicating if the optimizer exited successfully and
+            `message` which describes the cause of the termination. See
+            `OptimizeResult` for a description of other attributes.
+        """
+        for x, fun in self:
+            self.x, self.fun = x, fun
+
+            if (self.converged() or self.nfev >= self.maxfun or
+                    self.nit >= self.maxiter):
+                break
+
+        self._finish_up()
+        return self.result
+
+    def converged(self):
+        """
+        The truth of whether the solver has converged.
+        """
+        raise NotImplementedError
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """
+        Advance the solver by a single iteration.
+        """
+        # Should be over-ridden by each class based solver.
+        # TODO, Stop if self.maxfun/self.maxiter reached?
+        raise NotImplementedError
+    # step is because PyTorch has a step method
+    next = step = __next__
```

Comment:
We should move this to `__init__` and have `self.next = self.step = self.__next__`. That makes sure they all refer to the same object (at least when `super().__init__` called)

Reply:
Done